pam_console/pam_console_apply.8
.\" Copyright 2001 Red Hat, Inc.
.TH pam_console_apply 8 2005/5/2 "Red Hat" "System Administrator's Manual"
.SH NAME
pam_console_apply \- set or revoke permissions for users at the system console
.SH SYNOPSIS
.B pam_console_apply 
[-f <fstab file>] [-c <console.perms file>] [-r] [-t <tty>] [-s] [-d] [<device file> ...]
.SH DESCRIPTION
\fBpam_console_apply\fP is a helper executable which sets or resets permissions
on device nodes.
.br
If \fI/var/run/console.lock\fP exists, \fBpam_console_apply\fP will grant
permissions to the user listed therein.  If the lock file does not exist,
permissions are reset according to defaults set in \fIconsole.perms\fP files,
normally configured to set permissions on devices so that \fBroot\fP
owns them.

When initializing its configuration it first parses
the \fI/etc/security/console.perms\fP file and then it searches for files
ending with the \fI.perms\fP suffix in the \fI/etc/security/console.perms.d\fP
directory. These files are parsed in the lexical order in "C" locale.
Permission rules are appended to a global list, console and device class
definitions override previous definitions of the same class.
.SH ARGUMENTS
.IP -c
Load other console.perms file than the default one.
.IP -f
Load other fstab file than the default one (\fI/etc/fstab\fP).
.IP -r
Signals \fBpam_console_apply\fP to reset permissions.  The default is to set
permissions so that the user listed in \fI/var/run/console.lock\fP has access
to the devices, and to reset permissions if no such file exists.
.IP -t
Use <tty> to match console class in console.perms file. The default is tty0.
.IP -s
Write error messages to the system log instead of stderr.
.IP -d
Log/display messages useful for debugging.
.PP
The optional <device file> arguments constrain what files should be affected
by \fBpam_console_apply\fP. If they aren't specified permissions are
changed on all files specified in the \fIconsole.perms\fP file.
.SH FILES
\fI/var/run/console.lock\fP
.br
\fI/etc/security/console.perms\fP
.br
\fI/etc/security/console.perms.d/50-default.perms\fP
.SH "SEE ALSO"
.BR pam_console(8)
.br
.BR console.perms(5)
.br
.SH BUGS
Let's hope not, but if you find any, please report them via the "Bug Track"
link at http://bugzilla.redhat.com/bugzilla/
.SH AUTHORS
Nalin Dahyabhai <nalin@redhat.com>, using code shamelessly stolen from parts of
pam_console.
.br
Support of console.perms.d and other improvements by
Tomas Mraz <tmraz@redhat.com>.


pam_console/pam_console_apply.c
/*
 * Read in the file, and grant ownerships to whoever has the lock.
 */

#include "../../_pam_aconf.h"
#include <errno.h>
#include <glib.h>
#include <pwd.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <stdarg.h>
#include <glob.h>
#include <locale.h>
#define STATIC static
#include "chmod.h"
#include "pam_console.h"
#include "config.h"

#include <security/_pam_macros.h>

#define CAST_ME_HARDER (const void**)
#define DEFAULT_PERMSFILE "/etc/security/console.perms"
#define PERMS_GLOB "/etc/security/console.perms.d/*.perms"

static const char consolelock[] = LOCKDIR "/" LOCKFILE;
static char consoleperms[PATH_MAX];
static char tty[PATH_MAX] = "tty0";
static int debug = 0;
static int syslogging = 0;

static void *
_do_malloc(size_t req)
{
	void *ret;
	ret = malloc(req);
	if (!ret) abort();
	return ret;
}

static void
_pam_log(int err, int debug_p, const char *format, ...)
{
	va_list args;
	if (debug_p && !debug) return;
	va_start(args, format);
	if (syslogging) {
		openlog("pam_console_apply", LOG_CONS|LOG_PID, LOG_AUTHPRIV);
		vsyslog(err, format, args);
		closelog();
	}
	else {
		vfprintf(stderr, format, args);
		fprintf(stderr, "\n");
	}
	va_end(args);
}

static int
pf_glob_errorfn(const char *epath, int eerrno)
{
	return 0;
}

static void
parse_files(void)
{
	int rc;
	glob_t globbuf;
	int i;
	const char *oldlocale;

	/* first we parse the console.perms file */
	parse_file(DEFAULT_PERMSFILE);

	/* set the LC_COLLATE so the sorting order doesn't depend
	on system locale */
	oldlocale = setlocale(LC_COLLATE, "C");

	rc = glob(PERMS_GLOB, GLOB_NOCHECK, pf_glob_errorfn, &globbuf);
	setlocale(LC_COLLATE, oldlocale);
	if (rc == GLOB_NOSPACE) {
		return;
	}

	for (i = 0; globbuf.gl_pathv[i] != NULL; i++) {
		parse_file(globbuf.gl_pathv[i]);
	}
	globfree(&globbuf);
}

int
main(int argc, char **argv)
{
	int fd;
	int i, c;
	struct stat st;
	char *consoleuser = NULL;
	enum {Set, Reset} sense = Set;
	GSList *files = NULL;

	while((c = getopt(argc, argv, "c:f:t:rsd")) != -1) {
		switch(c) {
			case 'c': if (strlen(optarg) >= sizeof(consoleperms)) {
					fprintf(stderr, "Console.perms filename too long\n");
					exit(1);
				  }
				  strncpy(consoleperms, optarg, sizeof(consoleperms) - 1);
				  consoleperms[sizeof(consoleperms) - 1] = '\0';
				  break;
			case 'f': chmod_set_fstab(optarg);
				  break;
			case 't': if (strlen(optarg) >= sizeof(tty)) {
					fprintf(stderr, "TTY name too long\n");
					exit(1);
				  }
				  strncpy(tty, optarg, sizeof(tty) - 1);
				  tty[sizeof(tty) - 1] = '\0';
				  break;
			case 'r':
				  sense = Reset;
				  break;
			case 's': 
				  syslogging = TRUE;
				  break;
			case 'd': 
				  debug = TRUE;
				  break;
			default:
				  fprintf(stderr, "usage: %s [-f /etc/fstab] "
					  "[-c %s] [-t tty] [-r] [-s] [-d] [<device file> ...]\n", argv[0],
					  consoleperms);
				  exit(1);
		}
	}

	for (i = argc-1; i >= optind;  i--) {
		files = g_slist_prepend(files, argv[i]);
        }

	if (*consoleperms == '\0')
		parse_files();
	else
		parse_file(consoleperms);
		
        fd = open(consolelock, O_RDONLY);
	if (fd != -1) {
		if (fstat (fd, &st)) {
			_pam_log(LOG_ERR, FALSE,
			       "\"impossible\" fstat error on %s", consolelock);
			close(fd);
			goto return_error;
		}
		if (st.st_size) {
			consoleuser = _do_malloc(st.st_size+1);
			memset(consoleuser, '\0', st.st_size);
			if ((i = read (fd, consoleuser, st.st_size)) == -1) {
				_pam_log(LOG_ERR, FALSE,
				       "\"impossible\" read error on %s",
				       consolelock);
				goto return_error;
			}
			consoleuser[i] = '\0';
		}
		close(fd);
	} else {
		sense = Reset;
	}
	if((sense == Set) && (consoleuser != NULL)) {
		set_permissions(tty, consoleuser, files);
	}
	if(sense == Reset) {
		reset_permissions(tty, files);
	}
	return 0;

return_error:
	return 1;
}

/* supporting functions included from other .c files... */

#include "regerr.c"
#include "chmod.c"
#include "modechange.c"
#include "config.lex.c"
#include "config.tab.c"
